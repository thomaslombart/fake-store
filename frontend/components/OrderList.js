import React from "react";
import { useQuery } from "@apollo/react-hooks";
import Link from "next/link";
import gql from "graphql-tag";
import { distanceInWords } from "date-fns";

import Error from "./ErrorMessage";
import formatMoney from "../lib/formatMoney";

const USER_ORDERS_QUERY = gql`
  query USER_ORDERS_QUERY {
    orders(orderBy: createdAt_DESC) {
      id
      total
      createdAt
      items {
        id
        title
        price
        description
        quantity
        image
      }
    }
  }
`;

const OrderList = () => {
  const { loading, error, data } = useQuery(USER_ORDERS_QUERY);

  if (loading) return <p>loading...</p>;
  if (error) return <Error erorr={error} />;

  const orders = data.orders;

  return (
    <div>
      <h2>You have {orders.length} orders</h2>
      <ul>
        {orders.map(order => {
          return (
            <div key={order.id}>
              <Link
                href={{
                  pathname: "/order",
                  query: { id: order.id }
                }}
              >
                <a>
                  <div className="order-meta">
                    <p>
                      {order.items.reduce((a, b) => a + b.quantity, 0)} Items
                    </p>
                    <p>{order.items.length} Products</p>
                    <p>{distanceInWords(order.createdAt, new Date())}</p>
                    <p>{formatMoney(order.total)}</p>
                  </div>
                  <div className="images">
                    {order.items.map(item => (
                      <img key={item.id} src={item.image} alt={item.title} />
                    ))}
                  </div>
                </a>
              </Link>
            </div>
          );
        })}
      </ul>
    </div>
  );
};

export default OrderList;
export { USER_ORDERS_QUERY };
