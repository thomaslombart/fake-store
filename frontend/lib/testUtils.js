import casual from "casual";
import { render as rtl } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";
import { useApolloClient } from "@apollo/react-hooks";

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  __typename: "Item",
  id: "abc123",
  price: 5000,
  user: null,
  image: "cat-small.jpg",
  title: "cats are best",
  description: "cats",
  largeImage: "cat.jpg"
});

const secondFakeItem = () => ({
  __typename: "Item",
  id: "azerty",
  price: 3500,
  user: null,
  image: "dog-small.jpg",
  title: "dogs are best",
  description: "dogs",
  largeImage: "dog.jpg"
});

const fakeUser = () => ({
  __typename: "User",
  id: "4234",
  name: casual.name,
  email: casual.email,
  permissions: ["ADMIN"],
  cart: []
});

const fakeOrderItem = () => ({
  __typename: "OrderItem",
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words()
});

const fakeOrder = () => ({
  __typename: "Order",
  id: "ord123",
  charge: "ch_123",
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: "2018-04 - 06T19: 24: 16.000Z",
  user: fakeUser()
});

const fakeCartItem = overrides => ({
  __typename: "CartItem",
  id: "omg123",
  quantity: 3,
  item: fakeItem(),
  user: fakeUser(),
  ...overrides
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

const render = (component, mocks = []) => {
  let apolloClient;

  const Wrapper = () => {
    const client = useApolloClient();

    if (client) {
      apolloClient = client;
    } else {
      throw new Error("Apollo Client is not initialized.");
    }

    return <div>{component}</div>;
  };

  return {
    ...rtl(
      <MockedProvider mocks={mocks}>
        <Wrapper />
      </MockedProvider>
    ),
    apolloClient
  };
};

export {
  render,
  LocalStorageMock,
  fakeItem,
  secondFakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem
};
