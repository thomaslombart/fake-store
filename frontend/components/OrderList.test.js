import { waitForElement } from "@testing-library/react";

import { render, fakeOrder } from "../lib/testUtils";
import OrderList, { USER_ORDERS_QUERY } from "./OrderList";

const fakeOrders = [
  { ...fakeOrder(), id: "1" },
  { ...fakeOrder(), id: "2" },
  fakeOrder()
];

const mocks = [
  {
    request: {
      query: USER_ORDERS_QUERY
    },
    result: {
      data: {
        orders: fakeOrders
      }
    }
  }
];

describe("OrderList", () => {
  it("displays orders", async () => {
    const { getByText, baseElement } = render(<OrderList />, mocks);
    await waitForElement(() => getByText(/You have 3 orders/i));
    expect(baseElement).toMatchSnapshot();
  });
});
