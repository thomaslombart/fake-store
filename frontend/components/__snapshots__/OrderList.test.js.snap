// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderList displays orders 1`] = `
<body>
  <div>
    <div>
      <div>
        <h2>
          You have 
          3
           orders
        </h2>
        <ul>
          <div>
            <a
              href="/order?id=1"
            >
              <div
                class="order-meta"
              >
                <p>
                  2
                   Items
                </p>
                <p>
                  2
                   Products
                </p>
                <p>
                  over 1 year
                </p>
                <p>
                  $400
                </p>
              </div>
              <div
                class="images"
              >
                <img
                  alt="soluta non omnis consequatur enim quia autem"
                  src="reprehenderit.jpg"
                />
                <img
                  alt="quia sed exercitationem omnis laborum exercitationem est"
                  src="sint.jpg"
                />
              </div>
            </a>
          </div>
          <div>
            <a
              href="/order?id=2"
            >
              <div
                class="order-meta"
              >
                <p>
                  2
                   Items
                </p>
                <p>
                  2
                   Products
                </p>
                <p>
                  over 1 year
                </p>
                <p>
                  $400
                </p>
              </div>
              <div
                class="images"
              >
                <img
                  alt="nihil quam nulla est unde sit ex"
                  src="dolorem.jpg"
                />
                <img
                  alt="voluptas culpa et vero repellat eveniet assumenda"
                  src="et.jpg"
                />
              </div>
            </a>
          </div>
          <div>
            <a
              href="/order?id=ord123"
            >
              <div
                class="order-meta"
              >
                <p>
                  2
                   Items
                </p>
                <p>
                  2
                   Products
                </p>
                <p>
                  over 1 year
                </p>
                <p>
                  $400
                </p>
              </div>
              <div
                class="images"
              >
                <img
                  alt="odit tempora at modi occaecati illum laborum"
                  src="at.jpg"
                />
                <img
                  alt="possimus sint fugiat maiores autem est esse"
                  src="et.jpg"
                />
              </div>
            </a>
          </div>
        </ul>
      </div>
    </div>
  </div>
</body>
`;
